src/sessions.ts

```diff
@@ -189,21 +241,25 @@ export class ClientSession extends TypedEventEmitter<ClientSessionEvents> {
    */
   endSession(): Promise<void>;
   endSession(callback: Callback<void>): void;
-  endSession(options: Record<string, unknown>): Promise<void>;
-  endSession(options: Record<string, unknown>, callback: Callback<void>): void;
+  endSession(options: EndSessionOptions): Promise<void>;
+  endSession(options: EndSessionOptions, callback: Callback<void>): void;
   endSession(
-    options?: Record<string, unknown> | Callback<void>,
+    options?: EndSessionOptions | Callback<void>,
     callback?: Callback<void>
   ): void | Promise<void> {
     if (typeof options === 'function') (callback = options), (options = {});
-    options = options ?? {};
+    const _options = options ?? ({} as EndSessionOptions);
+    if (_options.force == null) _options.force = true;
 
     return maybePromise(callback, done => {
       if (this.hasEnded) {
+        maybeClearPinnedConnection(this, _options);
```

Comment:
do the new tests fully exercise this new pinning/unpinning logic? the code *seems* fine, but without it being clearly outlined in the tests (or elsewhere?) exactly when things are supposed to get pinned/unpinned, it's hard to tell by just looking at this file whether the resulting behavior will be 100% correct... case in point, that comment regarding unpinning on errors in L477; I'd be more comfortable with these changes if we had a single set of tests that outlined the expected behavior

Reply:
The load balancer spec tests add a new assertion to the unified test runner, `assertNumberConnectionsCheckedOut`, which is checking to make sure connections remained pinned or are returned to the pool based on the pool's count. These are tested for both cursors (https://github.com/mongodb/node-mongodb-native/blob/4.0/test/spec/load-balancers/cursors.json) and transactions (https://github.com/mongodb/node-mongodb-native/blob/4.0/test/spec/load-balancers/transactions.json)

Reply:
are they also tested for non load balanced clusters? and is there something that needs to happen to address the comment in L477?

Reply:
No there are no tests for that as connection pinning does not happen outside of load balancer mode and is only part of the LB test suite.

Reply:
I guess my confusion is more around the two types of pinning that we do, some of which is now handled by the same logic - there was already session server pinning and the session.transaction.unpinServer code that was in place... I'm trying to understand how the purpose of the original logic relates to the new connection pinning

Reply:
Yes the pinning logic was combined, just the resource getting pinned is different. But according to the transaction and load balancer specs, the pinning behaviour is the same, we are just pinning to a `Connection` in LB mode (to ensure executing against the same mongos behind the load balancer) and a `Server` in non-lb mode (since it's a direct connection to the mongos and any connection in the server's connection pool with execute on it)